File: workflows/src/utils/object_store.py
import os
from pathlib import Path
import shutil
from dataclasses import dataclass
from typing import Any, Dict, Optional, Tuple

import boto3
import fsspec
from botocore.config import Config


@dataclass(frozen=True)
class ObjectStoreConfig:
    endpoint: str
    region: str
    access_key: str
    secret_key: str
    force_path_style: bool
    bucket: str


def _normalize_endpoint(endpoint: str) -> str:
    if endpoint.startswith("http://") or endpoint.startswith("https://"):
        return endpoint
    return f"https://{endpoint}"


def _parse_bool(value: Optional[str]) -> bool:
    if value is None:
        return False
    return value.strip().lower() in {"1", "true", "yes", "y"}


def get_object_store_config() -> ObjectStoreConfig:
    """
    Load object store configuration from environment variables.
    """
    endpoint = os.getenv("OBJECT_STORE_ENDPOINT")
    region = os.getenv("OBJECT_STORE_REGION", "us-east-1")
    access_key = os.getenv("OBJECT_STORE_ACCESS_KEY_ID")
    secret_key = os.getenv("OBJECT_STORE_SECRET_KEY_ID")
    bucket = os.getenv("OBJECT_STORE_BUCKET_NAME")
    force_path_style = _parse_bool(os.getenv("OBJECT_STORE_FORCE_PATH_STYLE"))

    if not endpoint:
        raise ValueError("OBJECT_STORE_ENDPOINT is not configured.")
    if not access_key or not secret_key:
        raise ValueError("Object storage credentials are not configured.")
    if not bucket:
        raise ValueError("OBJECT_STORE_BUCKET_NAME is not configured.")

    return ObjectStoreConfig(
        endpoint=_normalize_endpoint(endpoint),
        region=region,
        access_key=access_key,
        secret_key=secret_key,
        force_path_style=force_path_style,
        bucket=bucket,
    )


def get_source_object_store_config() -> ObjectStoreConfig:
    """
    Load source object store configuration for reading the Zarr layer library.
    """
    endpoint = (
        os.getenv("SOURCE_OBJECT_STORE_ENDPOINT")
        or os.getenv("SOURCE_OBJECT_STORE_URL")
        or os.getenv("OBJECT_STORE_ENDPOINT")
        or os.getenv("OBJECT_STORE_URL")
    )
    region = os.getenv("SOURCE_OBJECT_STORE_REGION") or os.getenv(
        "OBJECT_STORE_REGION", "us-east-1"
    )
    access_key = os.getenv("SOURCE_OBJECT_STORE_ACCESS_KEY_ID") or os.getenv(
        "OBJECT_STORE_ACCESS_KEY_ID"
    )
    secret_key = os.getenv("SOURCE_OBJECT_STORE_SECRET_KEY_ID") or os.getenv(
        "OBJECT_STORE_SECRET_KEY_ID"
    )
    bucket = os.getenv("SOURCE_OBJECT_STORE_BUCKET_NAME") or os.getenv(
        "OBJECT_STORE_BUCKET_NAME"
    )
    force_path_style = _parse_bool(
        os.getenv("SOURCE_OBJECT_STORE_FORCE_PATH_STYLE")
        or os.getenv("OBJECT_STORE_FORCE_PATH_STYLE")
    )

    if not endpoint:
        raise ValueError("SOURCE_OBJECT_STORE_ENDPOINT is not configured.")
    if not access_key or not secret_key:
        raise ValueError("Source object storage credentials are not configured.")
    if not bucket:
        raise ValueError("SOURCE_OBJECT_STORE_BUCKET_NAME is not configured.")

    return ObjectStoreConfig(
        endpoint=_normalize_endpoint(endpoint),
        region=region,
        access_key=access_key,
        secret_key=secret_key,
        force_path_style=force_path_style,
        bucket=bucket,
    )


def get_object_store_client(config: ObjectStoreConfig):
    """
    Build a configured boto3 S3 client for the object store.
    """
    client_config = Config()

    if config.force_path_style:
        client_config = Config(s3={"addressing_style": "path"})

    return boto3.client(
        "s3",
        endpoint_url=config.endpoint,
        aws_access_key_id=config.access_key,
        aws_secret_access_key=config.secret_key,
        region_name=config.region,
        config=client_config,
    )


def get_source_zarr_store(zarr_path: Optional[str]):
    """
    Build an xarray-compatible Zarr store using SOURCE_OBJECT_STORE_* when the
    configured path points at object storage.
    """
    if not zarr_path:
        raise ValueError("ZARR_STORE_PATH is not configured.")

    if not _is_remote_zarr_path(zarr_path) and not os.getenv(
        "SOURCE_OBJECT_STORE_BUCKET_NAME"
    ):
        return zarr_path.strip().strip("\"'")

    config = get_source_object_store_config()
    key = build_object_key(_normalize_zarr_path(zarr_path, config.bucket))
    storage_options: Dict[str, Any] = {
        "key": config.access_key,
        "secret": config.secret_key,
        "client_kwargs": {
            "endpoint_url": config.endpoint,
            "region_name": config.region,
        },
    }

    if config.force_path_style:
        storage_options["config_kwargs"] = {"s3": {"addressing_style": "path"}}

    fs = fsspec.filesystem("s3", **storage_options)

    return fs.get_mapper(f"{config.bucket}/{key}")


def get_source_boundary_key() -> str:
    """
    Build the source object key for the BC boundary asset.
    """
    boundary_path = os.getenv("SOURCE_BC_BOUNDARY_PATH")

    if not boundary_path:
        raise ValueError("SOURCE_BC_BOUNDARY_PATH is not configured.")

    config = get_source_object_store_config()
    return build_object_key(_normalize_source_path(boundary_path, config.bucket))


def download_source_object(
    *,
    key: str,
    local_path: str,
) -> str:
    """
    Download a source object or object prefix from source object storage to local path.
    """
    if not key:
        raise ValueError("Source object key is required.")
    if not local_path:
        raise ValueError("local_path is required.")

    config = get_source_object_store_config()
    local_target = Path(local_path)
    if local_target.exists():
        if local_target.is_file():
            local_target.unlink()
        else:
            shutil.rmtree(local_target)

    local_target.parent.mkdir(parents=True, exist_ok=True)

    storage_options: Dict[str, Any] = {
        "key": config.access_key,
        "secret": config.secret_key,
        "client_kwargs": {
            "endpoint_url": config.endpoint,
            "region_name": config.region,
        },
    }

    if config.force_path_style:
        storage_options["config_kwargs"] = {"s3": {"addressing_style": "path"}}

    fs = fsspec.filesystem("s3", **storage_options)
    fs.get(f"{config.bucket}/{key}", str(local_target), recursive=True)

    if not local_target.exists():
        raise FileNotFoundError(f"Failed to download source object to {local_target}")

    return str(local_target)


def build_object_key(key: str) -> str:
    """
    Build a normalized object key.
    """
    return key.lstrip("/")


def _is_remote_zarr_path(zarr_path: str) -> bool:
    normalized = zarr_path.strip().strip("\"'")

    return (
        normalized.startswith("s3://")
        or normalized.startswith("http://")
        or normalized.startswith("https://")
    )


def _normalize_zarr_path(zarr_path: str, bucket: Optional[str] = None) -> str:
    normalized = zarr_path.strip().strip("\"'")

    if normalized.startswith("s3://"):
        without_scheme = normalized[len("s3://") :]
        parts = without_scheme.split("/", 1)
        if len(parts) == 1:
            return ""
        return parts[1].strip("/")

    if normalized.startswith("http://") or normalized.startswith("https://"):
        without_scheme = normalized.split("://", 1)[1]
        path_parts = without_scheme.split("/")[1:]
        if bucket and path_parts and path_parts[0] == bucket:
            path_parts = path_parts[1:]
        return "/".join(path_parts).strip("/")

    return _strip_bucket_prefix(normalized, bucket)


def _strip_bucket_prefix(path: str, bucket: Optional[str]) -> str:
    """
    Strip a plain <bucket>/ prefix from source paths when present.
    """
    normalized_bucket = bucket.strip("/") if bucket else ""
    if normalized_bucket and path.startswith(f"{normalized_bucket}/"):
        return path[len(normalized_bucket) + 1 :]
    return path.strip("/")


def _normalize_source_path(source_path: str, bucket: Optional[str] = None) -> str:
    return _normalize_zarr_path(source_path, bucket)


def make_uri(bucket: str, key: str) -> str:
    """
    Build an S3 URI from bucket/key.
    """
    return f"s3://{bucket}/{key}"


def parse_uri(uri: str) -> Tuple[str, str]:
    """
    Parse an S3 URI into bucket/key.
    """
    if not uri.startswith("s3://"):
        raise ValueError(f"Unsupported object store URI: {uri}")
    without_scheme = uri[len("s3://") :]
    parts = without_scheme.split("/", 1)
    if len(parts) != 2 or not parts[0] or not parts[1]:
        raise ValueError(f"Invalid object store URI: {uri}")
    return parts[0], parts[1]


def put_object(
    *,
    local_path: str,
    bucket: str,
    key: str,
    content_type: Optional[str] = None,
    metadata: Optional[Dict[str, str]] = None,
) -> Dict[str, Optional[str]]:
    """
    Upload a local file to object storage.
    """
    config = get_object_store_config()
    client = get_object_store_client(config)
    extra_args: Dict[str, Any] = {}

    if content_type:
        extra_args["ContentType"] = content_type
    if metadata:
        extra_args["Metadata"] = metadata

    client.upload_file(local_path, bucket, key, ExtraArgs=extra_args or None)

    head = client.head_object(Bucket=bucket, Key=key)
    return {"uri": make_uri(bucket, key), "etag": head.get("ETag")}


def get_object(
    *,
    bucket: str,
    key: str,
) -> Any:
    """
    Fetch an object from storage and return the response body stream.
    """
    config = get_object_store_config()
    client = get_object_store_client(config)
    response = client.get_object(Bucket=bucket, Key=key)
    return response["Body"]


def head_object(
    *,
    bucket: str,
    key: str,
) -> Dict[str, Any]:
    """
    Fetch object metadata from storage.
    """
    config = get_object_store_config()
    client = get_object_store_client(config)
    return client.head_object(Bucket=bucket, Key=key)


def download_object(
    *,
    bucket: str,
    key: str,
    local_path: str,
) -> str:
    """
    Download an object to a local file path.
    """
    config = get_object_store_config()
    client = get_object_store_client(config)
    client.download_file(bucket, key, local_path)
    return local_path

File: api/src/utils/object-store.ts
import {
  GetObjectCommand,
  GetObjectCommandOutput,
  HeadObjectCommand,
  HeadObjectCommandOutput,
  PutObjectCommand,
  PutObjectCommandOutput,
  S3Client
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { Readable } from 'stream';

export interface ObjectStoreConfig {
  endpoint: string;
  region: string;
  accessKeyId: string;
  secretAccessKey: string;
  bucket: string;
  prefix: string;
  forcePathStyle: boolean;
}

/**
 * Returns the configured object store settings.
 */
export const getObjectStoreConfig = (): ObjectStoreConfig => {
  const endpoint = process.env.OBJECT_STORE_ENDPOINT;
  const region = process.env.OBJECT_STORE_REGION || 'us-east-1';
  const accessKeyId = process.env.OBJECT_STORE_ACCESS_KEY_ID;
  const secretAccessKey = process.env.OBJECT_STORE_SECRET_KEY_ID;
  const bucket = process.env.OBJECT_STORE_BUCKET_NAME || '';
  const prefix = (process.env.OBJECT_STORE_PREFIX || '').replace(/^\/+|\/+$/g, '');
  const forcePathStyle = String(process.env.OBJECT_STORE_FORCE_PATH_STYLE || '').toLowerCase() === 'true';

  if (!endpoint) {
    throw new Error('OBJECT_STORE_ENDPOINT is not configured.');
  }
  if (!accessKeyId || !secretAccessKey) {
    throw new Error('Object store credentials are not configured.');
  }
  if (!bucket) {
    throw new Error('OBJECT_STORE_BUCKET_NAME is not configured.');
  }

  return {
    endpoint: normalizeEndpoint(endpoint),
    region,
    accessKeyId,
    secretAccessKey,
    bucket,
    prefix,
    forcePathStyle
  };
};

/**
 * Returns the public endpoint used by clients (e.g. localhost for MinIO in dev).
 */
export const getObjectStorePublicEndpoint = (): string | null => {
  const endpoint = process.env.OBJECT_STORE_PUBLIC_ENDPOINT;

  if (!endpoint) {
    return null;
  }

  return normalizeEndpoint(endpoint);
};

/**
 * Build an S3 client for the configured object store.
 */
export const getObjectStoreClient = (): S3Client => {
  const config = getObjectStoreConfig();
  return new S3Client({
    endpoint: config.endpoint,
    credentials: {
      accessKeyId: config.accessKeyId,
      secretAccessKey: config.secretAccessKey
    },
    forcePathStyle: config.forcePathStyle,
    region: config.region
  });
};

/**
 * Build an object key with an optional prefix.
 */
export const buildObjectKey = (key: string): string => {
  const { prefix } = getObjectStoreConfig();
  if (!prefix) {
    return key.replace(/^\/+/, '');
  }
  return `${prefix}/${key.replace(/^\/+/, '')}`;
};

/**
 * Build an S3 URI from a bucket and key.
 */
export const makeUri = (bucket: string, key: string): string => {
  return `s3://${bucket}/${key}`;
};

/**
 * Parse an S3 URI into bucket and key components.
 */
export const parseUri = (uri: string): { bucket: string; key: string } => {
  if (!uri.startsWith('s3://')) {
    throw new Error(`Unsupported object store URI: ${uri}`);
  }

  const withoutScheme = uri.slice('s3://'.length);
  const parts = withoutScheme.split('/');

  if (parts.length < 2 || !parts[0]) {
    throw new Error(`Invalid object store URI: ${uri}`);
  }

  return {
    bucket: parts[0],
    key: parts.slice(1).join('/')
  };
};

/**
 * Generate a presigned GET URL for an object in the configured object store.
 *
 * @param {string} bucket
 * @param {string} key
 * @param {number} [expiresInSeconds]
 * @return {*}  {Promise<string | null>}
 */
export const getPresignedObjectUrl = async (
  bucket: string,
  key: string,
  expiresInSeconds = 300
): Promise<string | null> => {
  if (!bucket || !key) {
    return null;
  }

  const client = getObjectStoreClient();

  const signedUrl = await getSignedUrl(
    client,
    new GetObjectCommand({
      Bucket: bucket,
      Key: key
    }),
    {
      expiresIn: expiresInSeconds
    }
  );

  return rewriteSignedUrlForPublicEndpoint(signedUrl);
};

/**
 * Rewrites a signed URL to use a public endpoint when configured.
 *
 * @param {string} signedUrl
 * @return {*}  {string}
 */
const rewriteSignedUrlForPublicEndpoint = (signedUrl: string): string => {
  const publicEndpoint = getObjectStorePublicEndpoint();

  if (!publicEndpoint) {
    return signedUrl;
  }

  try {
    const signed = new URL(signedUrl);
    const publicUrl = new URL(publicEndpoint);

    signed.protocol = publicUrl.protocol;
    signed.host = publicUrl.host;

    return signed.toString();
  } catch {
    return signedUrl;
  }
};

/**
 * Upload an object to the configured object store.
 */
export const putObject = async (options: {
  bucket: string;
  key: string;
  body: Buffer | Uint8Array | Readable | string;
  contentType?: string;
  metadata?: Record<string, string>;
}): Promise<{ uri: string; etag?: string }> => {
  const client = getObjectStoreClient();
  const response = await client.send(
    new PutObjectCommand({
      Bucket: options.bucket,
      Key: options.key,
      Body: options.body,
      ContentType: options.contentType,
      Metadata: options.metadata
    })
  );

  return {
    uri: makeUri(options.bucket, options.key),
    etag: response.ETag
  };
};

/**
 * Retrieve an object from the configured object store.
 */
export const getObject = async (options: {
  bucket: string;
  key: string;
}): Promise<GetObjectCommandOutput> => {
  const client = getObjectStoreClient();
  return client.send(
    new GetObjectCommand({
      Bucket: options.bucket,
      Key: options.key
    })
  );
};

/**
 * Retrieve object metadata from the configured object store.
 */
export const headObject = async (options: {
  bucket: string;
  key: string;
}): Promise<HeadObjectCommandOutput> => {
  const client = getObjectStoreClient();
  return client.send(
    new HeadObjectCommand({
      Bucket: options.bucket,
      Key: options.key
    })
  );
};

const normalizeEndpoint = (endpoint: string): string => {
  if (endpoint.startsWith('http://') || endpoint.startsWith('https://')) {
    return endpoint;
  }
  return `https://${endpoint}`;
};
